internal/server/data/model/generate.go

```diff
@@ -0,0 +1,234 @@
+package model
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/format"
+	"go/parser"
+	"go/token"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"sort"
+	"strings"
+)
+
+// GenerateMethods for a model struct. These methods allow the model to
+// perform queries against a database table without repeating the names of
+// columns and fields.
+func GenerateMethods(model any, columns []string, filename string) error {
+	rt := reflect.TypeOf(model)
+
+	var fileset token.FileSet
+	var mode = parser.ParseComments | parser.AllErrors
+	file, err := parser.ParseFile(&fileset, filename, nil, mode)
+	if err != nil {
+		return fmt.Errorf("failed to source file: %w", err)
+	}
+
+	recvName := strings.ToLower(rt.Name()[:1])
+	methodIndex := make(map[string]int)
+	for i, decl := range file.Decls {
+		fn, ok := decl.(*ast.FuncDecl)
+		if !ok || fn.Recv == nil {
+			continue
+		}
+		if astIdentForNode(fn.Recv.List[0].Type).Name != rt.Name() {
+			continue
+		}
+
+		if names := fn.Recv.List[0].Names; len(names) > 0 && names[0].Name != "_" {
+			recvName = names[0].Name
```

Comment:
Any way to test this file? I find the code really hard to follow personally, without some examples I can run through the test suite